dm/master/shardddl/pessimist_test.go

```diff
@@ -279,3 +287,518 @@ func (t *testPessimist) TestPessimist(c *C) {
+	c.Assert(len(opCh), Equals, 1)
+	op11 := <-opCh
+	c.Assert(op11.Exec, IsTrue)
+	c.Assert(op11.Done, IsFalse)
+
+	// 9. wait exec operation for the non-owner become available.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		opCh = make(chan pessimism.Operation, 10)
+		ctx2, cancel2 = context.WithTimeout(ctx, watchTimeout)
+		pessimism.WatchOperationPut(ctx2, etcdTestCli, task, source2, rev2, opCh)
+		cancel2()
+		close(opCh)
+		c.Assert(len(opCh), Equals, 1)
+		op := <-opCh
+		c.Assert(op.Exec, IsFalse)
+		c.Assert(op.Done, IsFalse)
+	}()
+
+	// 10. mark exec operation for the owner as `done` (and delete the info).
+	op11c := op11
+	op11c.Done = true
+	done, _, err := pessimism.PutOperationDeleteExistInfo(etcdTestCli, op11c, i11)
+	c.Assert(err, IsNil)
+	c.Assert(done, IsTrue)
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		return p.Locks()[ID].IsDone(source1)
+	}), IsTrue)
+	wg.Wait()
+
+	// 11. re-PUT i12, to simulate the re-entrant of the non-owner after the lock become synced.
+	rev2, err = pessimism.PutInfo(etcdTestCli, i12)
+	c.Assert(err, IsNil)
+
+	// 12. wait skip operation for the non-owner become available again (with new revision, without existing done).
+	opCh = make(chan pessimism.Operation, 10)
+	ctx2, cancel2 = context.WithTimeout(ctx, watchTimeout)
+	pessimism.WatchOperationPut(ctx2, etcdTestCli, task, source2, rev2, opCh)
+	cancel2()
+	close(opCh)
+	c.Assert(len(opCh), Equals, 1)
+	op12 := <-opCh
+	c.Assert(op12.Exec, IsFalse)
+	c.Assert(op12.Done, IsFalse)
+
+	// 13. mark skip operation for the non-owner as `done` (and delete the info).
+	op12c := op12
+	op12c.Done = true
+	done, _, err = pessimism.PutOperationDeleteExistInfo(etcdTestCli, op12c, i12)
+	c.Assert(err, IsNil)
+	c.Assert(done, IsTrue)
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		return p.Locks()[ID].IsDone(source2)
+	}), IsTrue)
+
+	// 14. re-PUT i13, to simulate the re-entrant of the owner after the lock become synced.
+	rev3, err := pessimism.PutInfo(etcdTestCli, i13)
+	c.Assert(err, IsNil)
+
+	// 15. wait skip operation for the non-owner become available again (with new revision, with existing done).
+	opCh = make(chan pessimism.Operation, 10)
+	ctx2, cancel2 = context.WithTimeout(ctx, watchTimeout)
+	pessimism.WatchOperationPut(ctx2, etcdTestCli, task, source3, rev3, opCh)
+	cancel2()
+	close(opCh)
+	c.Assert(len(opCh), Equals, 1)
+	op13 := <-opCh
+	c.Assert(op13.Exec, IsFalse)
+	c.Assert(op13.Done, IsFalse)
+
+	// 16. mark skip operation for the non-owner as `done` (and delete the info).
+	// the lock should become resolved now.
+	op13c := op13
+	op13c.Done = true
+	done, _, err = pessimism.PutOperationDeleteExistInfo(etcdTestCli, op13c, i13)
+	c.Assert(err, IsNil)
+	c.Assert(done, IsTrue)
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		_, ok := p.Locks()[ID]
+		return !ok
+	}), IsTrue)
+	t.noLockExist(c, p)
+}
+
+func (t *testPessimist) TestUnlockSourceLackBeforeSynced(c *C) {
```

Comment:
```suggestion
func (t *testPessimist) TestUnlockSourceLockBeforeSynced(c *C) {
```

Reply:
In fact, it is `Lack`, but I changed to use `Miss` in https://github.com/pingcap/dm/pull/522/commits/03479c5d62477e67593019b51810a60ded21f9d2.